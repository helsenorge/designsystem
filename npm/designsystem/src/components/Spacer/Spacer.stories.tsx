import React from 'react';

import { Meta, StoryObj } from '@storybook/react';

import Spacer from './Spacer';
import Docs from '../../docs';

const meta = {
  title: '@helsenorge/designsystem-react/Components/Spacer',
  component: Spacer,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={Spacer} />,
      description: {
        component:
          'Brukes til Ã¥ legge til luft i layout og mellom komponenter. For mer fleksibilitet med marginer og padding kan getSpacer funksjonen importeres og brukes fra scss ressursene.',
      },
    },
  },
} satisfies Meta<typeof Spacer>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: () => (
    <>
      <p>{'4xs'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'4xs'} />
      </div>
      <p>{'3xs'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'3xs'} />
      </div>
      <p>{'2xs'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'2xs'} />
      </div>
      <p>{'xs'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'xs'} />
      </div>
      <p>{'s'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'s'} />
      </div>
      <p>{'m'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'m'} />
      </div>
      <p>{'l'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'l'} />
      </div>
      <p>{'xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'xl'} />
      </div>
      <p>{'2xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'2xl'} />
      </div>
      <p>{'3xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'3xl'} />
      </div>
      <p>{'4xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'4xl'} />
      </div>
      <p>{'5xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'5xl'} />
      </div>
      <p>{'6xl'}</p>
      <div style={{ backgroundColor: 'lightgray', marginBottom: '0.25rem' }}>
        <Spacer size={'6xl'} />
      </div>
    </>
  ),
};
