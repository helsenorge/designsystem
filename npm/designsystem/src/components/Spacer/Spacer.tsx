import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';
import { SpacerSizes } from './../../theme/spacers';

import spacerStyles from './styles.module.scss';

export interface SpacerProps {
  /** Sets the size of the spacer. */
  size?: SpacerSizes;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

const Spacer = React.forwardRef(function SpacerForwardedRef(props: SpacerProps, ref: React.ForwardedRef<HTMLElement>) {
  const { size = 's', className, testId } = props;
  const spacerClasses = classNames(
    spacerStyles.spacer,
    {
      [spacerStyles['spacer--4xs']]: size === '4xs',
      [spacerStyles['spacer--3xs']]: size === '3xs',
      [spacerStyles['spacer--2xs']]: size === '2xs',
      [spacerStyles['spacer--xs']]: size === 'xs',
      [spacerStyles['spacer--s']]: size === 's',
      [spacerStyles['spacer--m']]: size === 'm',
      [spacerStyles['spacer--l']]: size === 'l',
      [spacerStyles['spacer--xl']]: size === 'xl',
      [spacerStyles['spacer--2xl']]: size === '2xl',
      [spacerStyles['spacer--3xl']]: size === '3xl',
      [spacerStyles['spacer--4xl']]: size === '4xl',
      [spacerStyles['spacer--5xl']]: size === '5xl',
      [spacerStyles['spacer--6xl']]: size === '6xl',
    },
    className
  );

  return <span className={spacerClasses} ref={ref} data-testid={testId} data-analyticsid={AnalyticsId.Spacer}></span>;
});

export default Spacer;
