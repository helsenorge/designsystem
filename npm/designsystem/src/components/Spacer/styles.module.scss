@import '../../scss/spacers';

.spacer {
  display: block;
  width: 100%;
  height: 0.125rem;

  &--4xs {
    height: getSpacer(4xs);
  }

  &--3xs {
    height: getSpacer(3xs);
  }

  &--2xs {
    height: getSpacer(2xs);
  }

  &--xs {
    height: getSpacer(xs);
  }

  &--s {
    height: getSpacer(s);
  }

  &--m {
    height: getSpacer(m);
  }

  &--l {
    height: getSpacer(l);
  }

  &--xl {
    height: getSpacer(xl);
  }

  &--2xl {
    height: getSpacer(2xl);
  }

  &--3xl {
    height: getSpacer(3xl);
  }

  &--4xl {
    height: getSpacer(4xl);
  }

  &--5xl {
    height: getSpacer(5xl);
  }

  &--6xl {
    height: getSpacer(6xl);
  }
}
