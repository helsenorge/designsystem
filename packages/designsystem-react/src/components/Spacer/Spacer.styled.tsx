import styled from 'styled-components';
import {SpacerSizes} from './../../theme/spacers';
import {getSpacer} from './../../theme/currys';

interface StyledSpacerProps {
  size: SpacerSizes;
}

const StyledSpacer = styled('span')<StyledSpacerProps>`
  display: block;
  width: 100%;
  height: ${props => getSpacer(props.size)};
`;

export default StyledSpacer;
