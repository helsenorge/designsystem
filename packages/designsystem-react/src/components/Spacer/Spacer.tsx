import React from 'react';
import StyledSpacer from './Spacer.styled';
import {SpacerSizes} from './../../theme/spacers';

interface SpacerProps {
  size?: SpacerSizes;
  className?: string;
}

const Spacer = React.forwardRef((props: SpacerProps, ref: any) => {
  const {size = 's', className} = props;

  return <StyledSpacer size={size} className={className} ref={ref}></StyledSpacer>;
});

export default Spacer;
