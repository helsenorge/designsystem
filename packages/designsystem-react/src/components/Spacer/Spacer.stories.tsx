import React from 'react';
import {withKnobs, text} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import Spacer from './Spacer';
import {withA11y} from '@storybook/addon-a11y';

const stories = storiesOf('Spacer', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{width: '20rem'}}>
    <p>{'3xs'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'3xs'} />
    </div>
    <p>{'2xs'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'2xs'} />
    </div>
    <p>{'xs'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'xs'} />
    </div>
    <p>{'s'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'s'} />
    </div>
    <p>{'m'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'m'} />
    </div>
    <p>{'l'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'l'} />
    </div>
    <p>{'xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'xl'} />
    </div>
    <p>{'2xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'2xl'} />
    </div>
    <p>{'3xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'3xl'} />
    </div>
    <p>{'4xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'4xl'} />
    </div>
    <p>{'5xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'5xl'} />
    </div>
    <p>{'6xl'}</p>
    <div style={{backgroundColor: 'lightgray', marginBottom: '0.25rem'}}>
      <Spacer size={'6xl'} />
    </div>
  </div>
));
