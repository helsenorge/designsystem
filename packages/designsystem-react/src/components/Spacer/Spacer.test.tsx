import React from 'react';
import {render} from '@testing-library/react';
import Spacer from './Spacer';

test('displays default spacer', (): void => {
  const {container} = render(<Spacer />);
  expect(container).toMatchSnapshot();
  expect(container.querySelector('span')).toHaveStyle('height: 1rem');
});

test('displays 3xs spacer', (): void => {
  const {container} = render(<Spacer size="3xs" />);
  expect(container.querySelector('span')).toHaveStyle('height: 0.125rem');
});

test('displays 2xs spacer', (): void => {
  const {container} = render(<Spacer size="2xs" />);
  expect(container.querySelector('span')).toHaveStyle('height: 0.25rem');
});

test('displays xs spacer', (): void => {
  const {container} = render(<Spacer size="xs" />);
  expect(container.querySelector('span')).toHaveStyle('height: 0.5rem');
});

test('displays s spacer', (): void => {
  const {container} = render(<Spacer size="s" />);
  expect(container.querySelector('span')).toHaveStyle('height: 1rem');
});

test('displays m spacer', (): void => {
  const {container} = render(<Spacer size="m" />);
  expect(container.querySelector('span')).toHaveStyle('height: 1.5rem');
});

test('displays l spacer', (): void => {
  const {container} = render(<Spacer size="l" />);
  expect(container.querySelector('span')).toHaveStyle('height: 2rem');
});

test('displays xl spacer', (): void => {
  const {container} = render(<Spacer size="xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 3rem');
});

test('displays 2xl spacer', (): void => {
  const {container} = render(<Spacer size="2xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 4rem');
});

test('displays 3xl spacer', (): void => {
  const {container} = render(<Spacer size="3xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 5rem');
});

test('displays 4xl spacer', (): void => {
  const {container} = render(<Spacer size="4xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 6rem');
});

test('displays 5xl spacer', (): void => {
  const {container} = render(<Spacer size="5xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 7rem');
});

test('displays 6xl spacer', (): void => {
  const {container} = render(<Spacer size="6xl" />);
  expect(container.querySelector('span')).toHaveStyle('height: 8rem');
});
