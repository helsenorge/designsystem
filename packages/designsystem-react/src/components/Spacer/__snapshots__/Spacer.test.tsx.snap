// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays default spacer 1`] = `
<div>
  <span
    class="sc-AxjAm fEkUGG"
  />
</div>
`;
